diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4610626..31f81fa 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,6 +2,8 @@ set(COMPONENT_SRCDIRS src
                       src/Fonts
                       src/utility)
 
+set(COMPONENT_SRCEXCLUDE src/M5StickC.cpp src/RTC.cpp)
+
 set(COMPONENT_ADD_INCLUDEDIRS src)
 
 set(COMPONENT_PRIV_INCLUDEDIRS src/Fonts 
@@ -13,3 +15,5 @@ set(COMPONENT_PRIV_INCLUDEDIRS src/Fonts
 set(COMPONENT_PRIV_REQUIRES arduino)
 
 register_component()
+
+target_compile_definitions(${COMPONENT_TARGET} PUBLIC "-DESP32")
\ No newline at end of file
